### DISCUSSION TITLE
StringSegment vs ReadOnlySpan<char> Implementation

### DISCUSSION THREAD

**Discussion by DepressionAlgorithm** [rds1983/FontStashSharp#52] [Ideas]
I apologize in advance for the verbosity and length of the post..

I like the new StringSegment ref struct in the newer version, a feature I was really wanting to see a lot. Though it occurred to me it's essentially the same as a ReadOnlySpan<char>, though ReadOnlySpan<T> may have different optimization (and in my testing, has very good performance).

Span<T> says it's a core2.1 feature, so I don't think core2.0 can use it. And I don't know why the project targets 2.0 rather than 2.1. If use of 2.0 over 2.1 is for compatibility with other libraries, switching to 2.1 would possibly break compat. If that's the case, updating to 2.1 and adding ReadOnlySpan<char> support may not be possible and you can ignore this idea (though down below I suggest char[] as a possible overload also).

I only mention it since, if I'm using a char[] array as the internal text buffer storage in my project (from a custom data structure, like a gap buffer for instance), there's no good way to convert from a ReadOnlySpan<char> (which converts implicitly from char[]) to StringSegment, since StringSegement expects a string in its constructor, and ReadOnlySpan<char> has no reference to the underlying array or string it's derived from.

Basically, StringSegment locks user into using 'string' as its internal data structure. Or, requires additional allocation of new strings by converting from the internal structure type to a fresh string copy, and then passing in the new string. This all but requires caching the strings, else you'd be allocating and de-allocating potentially many strings every rendering pass, so an entire additional caching layer would be needed between the internal buffer and SpriteFontBase. Doable, but possibly not ideal when a lot of text is involved, especially if the text is editable and constantly changing.

Changing over to ReadOnlySpan<char> has the benefit that it implicitly supports strings, char[] arrays and Memory<char>/ReadOnlyMemory<char>, where StringSegment only handles strings. ReadOnlySpan<char> gets a lot of additional flexibility.

Obviously StringBuilder is its own thing and requires special handling.

Possible solutions?

Switching from 2.0 to 2.1 is an obvious potential concern I assume.

Since you're using TextSource, adding ReadOnlySpan<char> would be quite easy. Just add the new ReadOnlySpan<char> overloads for the various API methods (Draw, Measure, etc). Add another constructor to TextSource, which is internal anyway so won't be bother end users at all, which accepts ReadOnlySpan<char>. Change StringSegment field to ReadOnlySpan<char>. StringSegment can be converted to ReadOnlySpan<char> with no addition allocations, as can a simple string of course. GetCodepoint can then concern itself with reading from the ReadOnlySpan<char>.

As an added benefit, char[] also implicitly converts to ReadOnlySpan<char>, so by adding a constructor overload for ReadOnlySpan<char>, you get implicit support for char[] for free.

The biggest drawback here is that StringSegment essentially becomes depricated, and maintaining support for it while adding support for ReadOnlySpan<char> does add overload bloat.

StringBuilder remains an exception case. Did I mention I hate strings and text? I hate dealing with strings and text.

To sum up, ReadOnlyMemory<char> can be easily added as method overload for various api methods(draw, measure, etc) and can be easily integrated into TextSource with little trouble.

Failing all this, or if switching to core2.1 isn't feasible, adding an overload to accept a char[], with a startIndex and Length, could be a possible way of adding a bit more flexibility. This would require adding a third field to TextSource though, one for a char[], one for StringSegment, and one for StringBuilder. This is needed because you can't convert from string to char[], or vice versa, without additional allocations and copying. So not ideal, but possible.

Either way, even if none of this is possible, the inclusion of indexable substring rendering is a welcome addition even if some string caching and conversion might be required. I realize this is approaching some fairly low-level optimization and many user's use case might not ever care or need this much flexibility, and usage of strings may be fine anyway.

I'm just obsessed with text rendering efficiency I guess.

I hate text.

I might have a problem.

Thanks for reading this wall of text crap if you made it this far. If you spot any mistakes let me know.

**Comment by rds1983** [comment]
FSS targets .net standard 2.0 in order to be compatible with [BRUTE](http://brute.rocks/), which is utility that is used to port games on platforms that doesnt have .NET, like Nintendo Switch. Problem is BRUTE requires a project to target NET452 and C# 5.0. Hence FSS has .net standard 2.0 limitation too. You can read more about that problem here: https://github.com/MonoGame/MonoGame/issues/7873

If you really need to draw part of char[], then it wouldn't be hard to add `ArraySegment<char>` to the TextSource as third possible source of chars. Probably eventually I'll even do it myself.

**Comment by DepressionAlgorithm** [comment]
Yeah I kind of figured it must be a reason like that. Unfortunate. But yeah, support for char[] would be nice since most custom text data structures are likely using a char[] anyways. I did download the repo months ago to add the Span overloads myself, but when I came back to my project a few days ago I noticed you had upgraded to core2.0 and added the StringSegment (which is close to what I was wanting anyway). I was hoping to avoid maintaining my own fork though.

Do you ever accept pull requests? I haven't done one before but I'd be willing to learn it and submit one since a change to implement `ArraySegment<char>` wouldn't be difficult, as you said. Just a few overloads for the SpriteFontBase methods and a tweak to TextSource as far as I can see.

`ArraySegment<char>` does have the limitation that it's not read-only since it exposes the underlying array (which is why they probably didn't add support for string I assume). It's possible user might not have a raw char[] as well. Either because it's a third party data structure which hides the raw array, or their custom structure hides the array and they don't want to write internal code to circumvent it.

Another option could be adding another custom ref struct type which accepts `IReadOnlyList<char>` along with offset/length. This interface is implemented by `char[]`, `ArraySegment<char>`, `List<char>` and various other char-based generics. It's also a dead simple interface to implement so user's custom data type could implement it easily. Or they could still just pass in the raw array if that's an option. They can all be stored in a `IReadOnlyList<char>` field and indexed, same as the StringSegment struct. There's a non-zero performance overhead for going through an interface rather than directly to an array of course, but I'm guessing it should be overall negligible. And this approach does maximize flexibility I think.

Sadly `string` does not support `IReadOnlyList<char>`. Apparently has to do with concerns over implementing .Count() and concerns about strings showing a bunch of LINQ extensions in intellisense. Shame, because the interface could be the binding type under the hood. As an added kick in the groin, StringBuilder also has an indexer and Length, meaning it could also have conceivably implemented `IReadOnlyList<char>` but likely doesn't for the same reasons as string.

Still, it would work just fine for char arrays and various other types, as well as any custom data structure which implements it, and it's a very common and nice interface.

Thoughts?
